### PR TITLE
[vm] Separation of resources, modules and aggregators

### PR DESCRIPTION
**Pull request by georgemitenkov**
### Description

TLDR; Cherrypicks some changes from #8227.

Motivation:
In Move VM we know what is a resource, what is a module. On top of
that adapter has aggregators. All these are merged in a single WriteSet.

Now, we split write sets into 3:
1. Resources
2. Modules
3. Aggregators

Why this is good:
1. Allows for better APIs in the future ti query different types, e.g. Move
    values or compiled module.
2. Aligns with `mvhashmap` having 3 different paths for resources, modules
    and aggregators.
3. No need to deserialize the state key to understand if this is code or not.
4. Aggregator write sets can be just `u128`, no need to have an actual WriteOp!
    (Note that this PR does not add this)
5. No need for WriteSetMut, etc. and other storage abstractions.

### CODE REVIEW

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 1]
```diff
@@ -60,71 +175,167 @@ impl VMChangeSet {
...
+            events,
+        };
+        checker.check_change_set(&change_set)?;
+        Ok(change_set)
+    }
+
+    /// Converts VM-native change set into its storage representation with fully
+    /// serialized changes.
+    /// At this points, if deltas are not materialized the conversion fails.
+    pub fn try_into_storage_change_set(self) -> anyhow::Result<StorageChangeSet, VMStatus> {
```
Comment (gelash) [inline aptos-move/aptos-vm-types/src/change_set.rs:220]: let's test this also?

**Review by gelash — APPROVED** [review]

File: aptos-move/aptos-aggregator/src/delta_change_set.rs [hunk 2]
```diff
@@ -333,6 +333,11 @@ impl DeltaChangeSet {
         self.delta_change_set.entry(key)
     }
 
+    #[inline]
+    pub fn contains(&self, key: &StateKey) -> bool {
```
Comment (ziaptos) [inline aptos-move/aptos-aggregator/src/delta_change_set.rs:337]: Why not keep naming consistent - `contains_key`

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 3]
```diff
@@ -6,51 +6,166 @@ use aptos_aggregator::delta_change_set::{deserialize, serialize, DeltaChangeSet}
...
+use std::collections::{
+    btree_map,
+    btree_map::{
+        Entry,
+        Entry::{Occupied, Vacant},
+    },
+    BTreeMap,
+};
+
+/// Encapsulates any changes to the global state.
```
Comment (ziaptos) [inline aptos-move/aptos-vm-types/src/change_set.rs:24]: Not really a meaningful comment. I'd just remove it

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 4]
```diff
@@ -60,71 +175,167 @@ impl VMChangeSet {
         Ok(change_set)
     }
 
-    pub fn write_set(&self) -> &WriteSet {
-        &self.write_set
+    /// Returns a new change set, built from the storage representation. Note
+    /// that this is a rather expensive operation because it re-splits resources
```
Comment (ziaptos) [inline aptos-move/aptos-vm-types/src/change_set.rs:179]: Great Warning about the complexity.  Maybe even rewrite such that the "WARNING" is clearly visible.

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 5]
```diff
@@ -60,71 +175,167 @@ impl VMChangeSet {
...
+    /// from modules and is only used to support transactions with write-set
+    /// payload.
+    pub fn try_from_storage_change_set(
```
Comment (ziaptos) [inline aptos-move/aptos-vm-types/src/change_set.rs:182]: nit:
try_ prefix and the return type seem redundant.

File: aptos-move/aptos-vm-types/src/change_set.rs (hunk 1 above)
Comment (ziaptos) [reply aptos-move/aptos-vm-types/src/change_set.rs:220]: [possibly] nit:
into -> to. unless it's used as a standard in Rust or in our codebase.

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 6]
```diff
@@ -60,71 +175,167 @@ impl VMChangeSet {
...
+        let write_set = write_set_mut.freeze().map_err(|_| {
+            VMStatus::error(
+                StatusCode::DATA_FORMAT_ERROR,
+                err_msg("Failed to freeze write sets when converting VMChangeSet to ChangeSet"),
+            )
+        })?;
+        Ok(StorageChangeSet::new(write_set, events))
+    }
+
+    /// A useful method to iterate over all writes.
```
Comment (ziaptos) [inline aptos-move/aptos-vm-types/src/change_set.rs:260]: nit: remove :)
all methods should be `Useful`, and the name clearly conveys what the method does.

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 7]
```diff
@@ -60,71 +175,167 @@ impl VMChangeSet {
...
+    }
+
+    /// Returns all changes made to aggregators.
+    pub fn aggregator_write_set(&self) -> &StateChange<StateKey, WriteOp> {
+        &self.aggregator_write_set
     }
 
+    /// Returns all the pending changes (i.e. deltas) made to aggregators
+    /// and not yet applied.
     pub fn delta_change_set(&self) -> &DeltaChangeSet {
```
Comment (ziaptos) [inline aptos-move/aptos-vm-types/src/change_set.rs:288]: 4 comments above are redundant with the function names.

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 8]
```diff
@@ -204,16 +405,69 @@ impl VMChangeSet {
...
-        let write_set = write_set_mut.freeze().map_err(|_| {
-            VMStatus::error(
-                StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
-                err_msg("Error when freezing squashed write sets."),
-            )
-        })?;
+        Ok(())
+    }
+
+    /// Squashes `next` writes on top of the write set.
```
Comment (ziaptos) [inline aptos-move/aptos-vm-types/src/change_set.rs:411]: naming:
`next` -> additional
`squash_writes` -> `squash_additional_writes`. 

And the comment becomes unnecessary.

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 9]
```diff
@@ -204,16 +405,69 @@ impl VMChangeSet {
...
+                },
+                Vacant(entry) => {
+                    entry.insert(next_write_op);
+                },
+            }
+        }
+        Ok(())
+    }
+
+    /// Squashes `next` change set on top of this change set. The squashed
```
Comment (ziaptos) [inline aptos-move/aptos-vm-types/src/change_set.rs:429]: see above

File: aptos-move/aptos-vm-types/src/output.rs [hunk 10]
```diff
@@ -2,27 +2,29 @@
...
 
 /// Output produced by the VM. Before VMOutput is passed to storage backends,
 /// it must be converted to TransactionOutput.
 #[derive(Debug, Clone)]
 pub struct VMOutput {
+    // All changes to the state, including data, code, events.
     change_set: VMChangeSet,
+    // Encapsulates all gas charges, e.g. execution, I/), storage, etc.
     fee_statement: FeeStatement,
+    // Status of the executed transaction.
```
Comment (ziaptos) [inline aptos-move/aptos-vm-types/src/output.rs:22]: 3 redundant comments [variable names and types convey that info]

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 11]
```diff
@@ -6,51 +6,166 @@ use aptos_aggregator::delta_change_set::{deserialize, serialize, DeltaChangeSet}
...
+    btree_map::{
+        Entry,
+        Entry::{Occupied, Vacant},
+    },
+    BTreeMap,
+};
+
+/// Encapsulates any changes to the global state.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct StateChange<K: PartialOrd + Ord, V>(BTreeMap<K, V>);
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm-types/src/change_set.rs:26]: Indeed!

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 12]
```diff
@@ -6,51 +6,166 @@ use aptos_aggregator::delta_change_set::{deserialize, serialize, DeltaChangeSet}
...
 pub struct VMChangeSet {
-    write_set: WriteSet,
+    // Changes to the data in the global state.
+    resource_write_set: StateChange<StateKey, WriteOp>,
+    // Changes to the code in the global state.
+    module_write_set: StateChange<StateKey, WriteOp>,
+    // Aggregator changes: writes and deltas.
+    // TODO(aggregator): For Aggregator V2, we will not use state keys
+    // (both for writes and deltas)!
+    aggregator_write_set: StateChange<StateKey, WriteOp>,
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm-types/src/change_set.rs:109]: Yes yes yes! I wanted to get rid of delta change set as well, through that `StateChange` thing though, and didn't do it here. With BTreeMaps the change is minimal though, let's do it

**Review by ziaptos — COMMENTED** [review]
Minor contribution from me from now (as I don't know a lot of background details).

A general comment is to try to avoid unnecessary and redundant comments - that makes maintaining the code harder in the future (one has to change both)

File: aptos-move/aptos-aggregator/src/delta_change_set.rs (hunk 2 above)
Comment (georgemitenkov) [reply aptos-move/aptos-aggregator/src/delta_change_set.rs:337]: Good point, actually got rid of StateChange thing

File: aptos-move/aptos-vm-types/src/change_set.rs (hunk 5 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm-types/src/change_set.rs:182]: Return result because there is this line:
```
checker.check_change_set(&change_set)?;
```
which checks the size of writes / bytes and errors if check does not pass.

File: aptos-move/aptos-vm-types/src/change_set.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm-types/src/change_set.rs:109]: I think let's keep writes and deltas separate for aggregator right now, it simplifies lots of logic in the PR and is slightly faster.

But also, we should revisit later. `aptos-aggregator` can be simplified a lot tbh, so we can do it then. Will have a look at this next.

File: aptos-move/aptos-vm-types/src/change_set.rs (hunk 1 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm-types/src/change_set.rs:220]: I think Rust uses `into` when consuming a value.

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 13]
```diff
@@ -170,20 +381,10 @@ impl VMChangeSet {
...
-                            err_msg(format!("Error while squashing two write ops: {}.", e)),
-                        )
-                    })?;
-                    if noop {
-                        entry.remove();
-                    }
+                    squash_write_ops!(entry, next_write_op);
                 },
                 Vacant(entry) => {
                     // This is a new write op. It can overwrite a delta so we
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm-types/src/change_set.rs:390]: Added more tests for that :)

File: aptos-move/aptos-vm-types/src/change_set.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (igor-aptos) [reply aptos-move/aptos-vm-types/src/change_set.rs:109]: I had the same comment. I'm fine for keeping it separate for this PR, but should merge all writes to an aggregator together in a single map after some of the other changes

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 14]
```diff
@@ -6,51 +6,166 @@ use aptos_aggregator::delta_change_set::{deserialize, serialize, DeltaChangeSet}
...
+    // (both for writes and deltas)!
+    aggregator_write_set: StateChange<StateKey, WriteOp>,
     delta_change_set: DeltaChangeSet,
+    // Events produced during the execution of a transaction.
     events: Vec<ContractEvent>,
 }
 
+// Useful macro sto squash a pair of write ops and ignore the result if
+// it is a no-op.
+macro_rules! squash_write_ops {
```
Comment (igor-aptos) [inline aptos-move/aptos-vm-types/src/change_set.rs:117]: just wondering, why is this a macro, not a function?

File: aptos-move/aptos-vm-types/src/change_set.rs (hunk 5 above)
Comment (igor-aptos) [inline aptos-move/aptos-vm-types/src/change_set.rs:182]: Where is try_from_storage_change_set used in the flow logically? Is this just for tests?

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 15]
```diff
@@ -60,71 +175,167 @@ impl VMChangeSet {
...
+
+        // Go over write ops to split them into resources and modules.
+        for (state_key, write_op) in write_set {
+            if matches!(state_key.inner(), StateKeyInner::AccessPath(ap) if ap.is_code()) {
+                module_write_set.insert(state_key, write_op);
+            } else {
+                // Everything else must be a resource.
+                // TODO(aggregator): Aggregator V1 is a table item, and so fits into
+                // this category. In practice this should never happen, but we might
+                // want to have an assert here before Aggregator V2 lands.
```
Comment (igor-aptos) [inline aptos-move/aptos-vm-types/src/change_set.rs:201]: hmm, what does go into aggregator_write_set, if not deltas and not writes to v1 aggregator table item?

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 16]
```diff
@@ -60,71 +175,167 @@ impl VMChangeSet {
...
+        // Create a write set which has all the changes combined together.
+        let mut write_set_mut = WriteSetMut::default();
+        write_set_mut.extend(resource_write_set);
+        write_set_mut.extend(module_write_set);
+
+        // TODO(aggregator): Aggregator V1 is a state item, and has to be added
+        // here.
+        // When we move to different key representation for aggregators,
+        // this has to be conditional on the table item (supply) until we
+        // fully transition to the new version.
```
Comment (igor-aptos) [inline aptos-move/aptos-vm-types/src/change_set.rs:248]: shouldn't after changes for v2 and ephemeral, v1 and v2 be handled the same?

i.e. v1 will have a statekey in resource_write_set (which only gets populated after materialization), and will have ephemeral change in the aggregator_write_set until materialization? (same as v2 aggregators?)

File: aptos-move/aptos-vm-types/src/output.rs [hunk 17]
```diff
@@ -126,24 +126,28 @@ impl VMOutput {
...
         // Add the delta writes to the write set of the transaction.
-        delta_writes.into_iter().for_each(|item| {
-            debug_assert!(
-                delta_change_set.remove(&item.0).is_some(),
-                "Delta writes contain a key which does not exist in DeltaChangeSet."
-            );
-            write_set_mut.insert(item)
-        });
-
+        let mut write_set_mut = WriteSetMut::new(delta_writes);
```
Comment (igor-aptos) [inline aptos-move/aptos-vm-types/src/output.rs:140]: this is the part that will become much more complicated in v2, i.e. go through deserialization/serialization path to patch values?

File: aptos-move/block-executor/src/executor.rs [hunk 18]
```diff
@@ -136,16 +136,32 @@ where
...
             }
 
-            // Then, apply deltas.
+            // Aggregators also require special handling: both rites and deltas go to the same map.
+            for (k, v) in output.get_aggregator_writes().into_iter() {
+                if !prev_modified_keys.remove(&k) {
+                    updates_outside = true;
+                }
+                versioned_cache.write(k, write_version, v);
+            }
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/executor.rs:164]: am I blind, I don't see any differences between these three loops? could we just chain the iterators and have one loop?

**Review by igor-aptos — CHANGES_REQUESTED** [review]
there are bunch of comments marked resolved, but code not yet changed - do you have some pending code locally to update the PR with? 
requesting changes for you to update the PR with that code.

File: aptos-move/block-executor/src/executor.rs (hunk 18 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/executor.rs:165]: That's correct, they are identical. We use the same mvhashmap right now, but there will be three -- one for each loop, so duplication will be removed very soon? I had a comment above about that. Let me know if you prefer chaining

File: aptos-move/aptos-vm-types/src/change_set.rs (hunk 5 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm-types/src/change_set.rs:87]: For Waypoint transaction, which carries a change set.

File: aptos-move/aptos-vm-types/src/change_set.rs (hunk 14 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm-types/src/change_set.rs]: Used to avoid `mut`s, etc., since the code is fully shared between two implementations. No strong preference, can make it a function

File: aptos-move/aptos-vm-types/src/change_set.rs (hunk 15 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm-types/src/change_set.rs]: Nothing? We construct from storage here, there are no aggregators there

File: aptos-move/aptos-vm-types/src/change_set.rs (hunk 16 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm-types/src/change_set.rs]: Yes, old comment, updated.

**Comment by georgemitenkov** [comment]
@igor-aptos PR updated

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 19]
```diff
@@ -60,71 +175,167 @@ impl VMChangeSet {
...
+                err_msg("Failed to freeze write sets when converting VMChangeSet to ChangeSet"),
+            )
+        })?;
+        Ok(StorageChangeSet::new(write_set, events))
+    }
+
+    /// A useful method to iterate over all writes.
+    /// TODO(aggregator): With Aggregator V2, we would have to revisit this
+    /// because we no longer iterate over state items.
+    pub fn write_set_iter(&self) -> impl Iterator<Item = (&StateKey, &WriteOp)> {
```
[the remark this answers is not in the document]
Comment (zekun000) [reply aptos-move/aptos-vm-types/src/change_set.rs:155]: looks beautiful

File: aptos-move/aptos-vm-types/src/change_set.rs [hunk 20]
```diff
@@ -151,69 +258,108 @@ impl VMChangeSet {
...
+            aggregator_delta_set: additional_aggregator_delta_set,
+            events: additional_events,
+        } = additional_change_set;
+        let Self {
+            mut resource_write_set,
+            mut module_write_set,
+            mut aggregator_write_set,
+            mut aggregator_delta_set,
+            mut events,
+        } = self;
```
Comment (zekun000) [inline aptos-move/aptos-vm-types/src/change_set.rs:344]: why do we need to unpack here? because we can't pass two &mut in the squash function?

File: aptos-move/aptos-vm/src/move_vm_ext/respawned_session.rs [hunk 21]
```diff
@@ -98,13 +100,31 @@ impl<'r> TStateView for ChangeSetStateView<'r> {
...
+            // The value to resolve delta has to come from `base`.
             Some(delta_op) => Ok(delta_op
                 .try_into_write_op(self.base, state_key)?
                 .as_state_value()),
-            None => match self.change_set.write_set().get(state_key) {
-                Some(write_op) => Ok(write_op.as_state_value()),
-                None => self.base.get_state_value(state_key),
+            None => {
+                // Otherwise, this can be either a resource or an aggregator write.
+                match self.change_set.resource_write_set().get(state_key) {
```
Comment (zekun000) [inline aptos-move/aptos-vm/src/move_vm_ext/respawned_session.rs:120]: this part is ugly, should we just have a get function for the change_set to check all the write ops?

File: aptos-move/block-executor/src/executor.rs (hunk 18 above)
Comment (zekun000) [reply aptos-move/block-executor/src/executor.rs:165]: i thought we already expose a chaining functions, we can just use that for now

**Review by igor-aptos — APPROVED** [review]

**Review by vgao1996 — APPROVED** [review]

File: aptos-move/aptos-vm/src/move_vm_ext/respawned_session.rs (hunk 21 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/move_vm_ext/respawned_session.rs:120]: Yes! I changed it to use write set iterator + find() for now
